sw/device/lib/crypto/impl/kdf.c

```diff
@@ -28,8 +29,109 @@ otcrypto_status_t otcrypto_kdf_kmac(
     const otcrypto_const_byte_buf_t kdf_context, size_t required_byte_len,
     otcrypto_blinded_key_t *keying_material) {
-  // TODO: Implement KMAC-KDF.
-  return OTCRYPTO_NOT_IMPLEMENTED;
+  // Check NULL pointers.
+  if (key_derivation_key.keyblob == NULL || keying_material == NULL) {
+    return OTCRYPTO_BAD_ARGS;
+  }
+
+  // Check the private key checksum.
+  if (integrity_blinded_key_check(&key_derivation_key) != kHardenedBoolTrue) {
+    return OTCRYPTO_BAD_ARGS;
+  }
+
+  // The underlying KMAC implementation is not hardened yet.
+  if (key_derivation_key.config.security_level !=
+      kOtcryptoKeySecurityLevelLow) {
+    return OTCRYPTO_NOT_IMPLEMENTED;
+  }
+
+  // Ensure `key_mode` is either KMAC-128 or KMAC-256.
+  if (key_derivation_key.config.key_mode != kOtcryptoKeyModeKdfKmac128 &&
+      key_derivation_key.config.key_mode != kOtcryptoKeyModeKdfKmac256) {
+    return OTCRYPTO_BAD_ARGS;
+  }
+
+  // Check if `key_mode` values match the input arg.
+  if (!((key_derivation_key.config.key_mode == kOtcryptoKeyModeKdfKmac128 &&
+         kmac_mode == kOtcryptoKmacModeKmac128) ||
+        (key_derivation_key.config.key_mode == kOtcryptoKeyModeKdfKmac256 &&
+         kmac_mode == kOtcryptoKmacModeKmac256))) {
+    return OTCRYPTO_BAD_ARGS;
+  }
+
+  // Validate key length of `key_derivation_key`.
+  if (key_derivation_key.config.hw_backed == kHardenedBoolTrue) {
+    return OTCRYPTO_NOT_IMPLEMENTED;
+  } else if (key_derivation_key.config.hw_backed == kHardenedBoolFalse) {
+    // Force `key_length` to be larger than minimum supported security level (
+    // 128-bit = 16 bytes).
+    if (key_derivation_key.config.key_length < 128 / 8) {
+      return OTCRYPTO_BAD_ARGS;
+    }
+    if (2 * key_derivation_key.config.key_length !=
+        key_derivation_key.keyblob_length) {
+      return OTCRYPTO_BAD_ARGS;
+    }
+  } else {
+    return OTCRYPTO_BAD_ARGS;
+  }
+
+  // Because of KMAC HWIPs prefix limitation, `label` should not exceed
+  // `kKmacCustStrMaxSize` bytes.
+  if (kdf_label.len > kKmacCustStrMaxSize) {
+    return OTCRYPTO_BAD_ARGS;
+  }
+
+  // Check non-zero length for keying_material.
+  if (required_byte_len == 0) {
+    return OTCRYPTO_BAD_ARGS;
+  }
+
+  // Check `keying_material` key length.
+  if (keying_material->config.hw_backed == kHardenedBoolTrue) {
+    return OTCRYPTO_NOT_IMPLEMENTED;
+  } else if (key_derivation_key.config.hw_backed == kHardenedBoolFalse) {
+    if (keying_material->config.key_length != required_byte_len ||
+        2 * required_byte_len != keying_material->keyblob_length) {
+      return OTCRYPTO_BAD_ARGS;
+    }
+    if (keying_material->keyblob == NULL) {
+      return OTCRYPTO_BAD_ARGS;
+    }
+  } else {
+    return OTCRYPTO_BAD_ARGS;
+  }
+
+  // The underlying KMAC implementation is not hardened yet.
+  if (keying_material->config.security_level != kOtcryptoKeySecurityLevelLow) {
+    return OTCRYPTO_NOT_IMPLEMENTED;
+  }
```

Comment:
nit: Maybe this check could move to be next to the same check for `key_derivation_key`, just so they're more logically grouped?